### PR TITLE
Add method to rotate a relative location around a location in the world.

### PR DESCRIPTION
**Pull request by ZeroErrors**
Use method for entity follow, still requires some tweaks and additions.

On line 67 do you want for example only the yaw to be used if hologram.isRelativePitch() is false and vice versa?
Also not sure what isRelativePitchControlMoreThanHeight is supposed to do so I just added some temp code in there.

Here is the plugin I used to test it:

``` Java
    @Override
    public void onEnable() {
        getServer().getPluginManager().registerEvents(this, this);
    }

    @EventHandler
    public void onJoin(PlayerJoinEvent event) {
        // X = Right(Neg) - Left(Pos)
        // Y = Up(Pos) - Down(Neg)
        // Z = Behind entity(Neg) - In front entity(Pos).
        new Hologram(event.getPlayer().getEyeLocation()
                .add(0, 0, 6), ChatColor.GOLD + "This is a hologram",
                ChatColor.RED + "Center").setFollowEntity(event.getPlayer(), false, true, true, false)
                .start();

        new Hologram(event.getPlayer().getEyeLocation()
                .add(0, 2, 6), ChatColor.GOLD + "Top").setFollowEntity(event.getPlayer(), false, true, true, false)
                .start();

        new Hologram(event.getPlayer().getEyeLocation()
                .add(0, -2, 6), ChatColor.GOLD + "Bottom").setFollowEntity(event.getPlayer(), false, true, true, false)
                .start();

        new Hologram(event.getPlayer().getEyeLocation()
                .add(-3, 0, 6), ChatColor.GOLD + "Right").setFollowEntity(event.getPlayer(), false, true, true, false)
                .start();

        new Hologram(event.getPlayer().getEyeLocation()
                .add(3, 0, 6), ChatColor.GOLD + "Left").setFollowEntity(event.getPlayer(), false, true, true, false)
                .start();
    }
```

### CODE REVIEW

**Comment by libraryaddict** [comment]
Mostly works. However what I use is
`event.getPlayer().getEyeLocation()
                .add(event.getPlayer().getEyeLocation().getDirection().multiply(10))`

This spawns the hologram 10 blocks in front of him at his cursor.

When using your commit, the hologram yaw isn't taken into account and so the hologram appears at the wrong yaw.

I use the cursor as the hologram point as its the best way to make sure its accurate.

I did try doing the yaw myself. But I failed.

**Comment by ZeroErrors** [comment]
What do you mean by hologram yaw?

**Comment by libraryaddict** [comment]
Use this

`new Hologram(event.getPlayer().getEyeLocation()
                .add(event.getPlayer().getEyeLocation().getDirection().multiply(10)), ChatColor.GOLD + "This is a hologram",
                ChatColor.RED + "This is the second line of the hologram", ChatColor.BLUE
                        + "This is the third line of the hologram").setFollowEntity(event.getPlayer(), false,false,true, false)
                .start();`

**Comment by libraryaddict** [comment]
It seems that when the location is set. I want to get the yaw offset when its set relative to the entity.

Then add it to the hologram when the entity moves.

So if its set at the players cursor. Then the offset is 0.
If its set 90 degrees to the players side.

Then the offset is 90.

Behind him is 180

**Comment by libraryaddict** [comment]
Also seeing as this is kind of more work than I originally thought it would be.

How does $30 sound?

**Comment by ZeroErrors** [comment]
I was just doing it more for fun than for money I was not going to take the $10.
I'll mess with it some more, is it ok to change some method names and maybe add in some other features that may help?

**Comment by libraryaddict** [comment]
Sure

**Comment by ZeroErrors** [comment]
Ok, So.
Instead of using:

``` Java
new Hologram(event.getPlayer().getEyeLocation() .add(event.getPlayer().getEyeLocation().getDirection().multiply(10)), ChatColor.GOLD + "This is a hologram", ChatColor.RED + "This is the second line of the hologram", ChatColor.BLUE + "This is the third line of the hologram").setFollowEntity(event.getPlayer(), false,false,true, false) .start();
```

try this:

``` Java
new Hologram(event.getPlayer().getEyeLocation() .add(0, 0, 10), ChatColor.GOLD + "This is a hologram", ChatColor.RED + "This is the second line of the hologram", ChatColor.BLUE + "This is the third line of the hologram").setFollowEntity(event.getPlayer(), false,false,true, false) .start();
```

I think to have this make more sense I will need probable need make a couple changes.

**Comment by libraryaddict** [comment]
That's not the problem however.

People would rather use the location they can understand, than make up a new location with lots of testing to make sure its the one they can understand.

For instance, I want to spawn a hologram in front of someones eyes and make it always stay there.

My first instinct is to use bukkits api and get the location in front of their eyes.

Its the easiest one to understand, rather than make people figure out what location they want.

**Comment by libraryaddict** [comment]
https://github.com/libraryaddict/Holograms/commit/71e0097d0607f24d1c349d553c9128fa8bd11e7d

This does what I want. However. It seems that your method is better.. When looking straight down, my version just puts the hologram straight down. Whereas your version will correct it and put it at the correct location in most cases.

So really. The sphere translation code is the problem.
When attempting to use your one, along with my yaw modifications. It just doesn't work.

**Comment by libraryaddict** [comment]
Heh. I finally understood what you were trying to get at with the non-relative pitch and yaw.

So really. It needs a option where it can do my method as well for those who want it.
My method wouldn't really be used much.

Can't believe I spent so long thinking you didn't understand 
_facepalms_

File: src/me/libraryaddict/holograms/HologramCentral.java [hunk 1]
```diff
@@ -224,4 +229,19 @@ public void onWorldSwitch(PlayerChangedWorldEvent event) {
...
+
+    private static Location rotateLocation(Location origin, Location relativeLocation, float yaw, float pitch) {
+        double yawRadians = Math.toRadians(-yaw);
+        double pitchRadians = Math.toRadians(pitch);
+
+        double rx = relativeLocation.getX();
+        double ry = relativeLocation.getY();
+        double rz = relativeLocation.getZ();
+
+        double x = rx * Math.cos(yawRadians) + (ry * Math.sin(pitchRadians) + rz * Math.cos(pitchRadians)) * Math.sin(yawRadians);
```
Comment (md-5) [inline src/me/libraryaddict/holograms/HologramCentral.java:241]: I'm sorry, this math is just wat. I don't even understand it.
https://github.com/md-5/SmallPlugins/blob/master/TestPlugin/src/main/java/net/md_5/TestPlugin.java

That is the correct and simple way of doing this.

Comment (ZeroErrors) [inline src/me/libraryaddict/holograms/HologramCentral.java:241]: Your code gets a point on the surface of the sphere, what this does is rotates a point around the origin.

Comment (md-5) [inline src/me/libraryaddict/holograms/HologramCentral.java:241]: Thats just a bit of addition / subtraction + Pythagoras, no need for a dozen more trig operations.

Comment (ZeroErrors) [inline src/me/libraryaddict/holograms/HologramCentral.java:241]: Well your welcome to make your own PR with a better way of doing it however this has been working just fine so far.
